crates/engine/tree/src/tree/mod.rs

```diff
@@ -155,9 +157,9 @@ impl TreeState {
 /// Container type for in memory state data.
 #[derive(Debug, Default)]
 pub struct InMemoryStateImpl {
-    blocks: HashMap<B256, Arc<State>>,
-    numbers: HashMap<u64, B256>,
-    pending: Option<State>,
+    blocks: RwLock<HashMap<B256, Arc<State>>>,
+    numbers: RwLock<HashMap<u64, B256>>,
+    pending: RwLock<Option<State>>,
```

Comment:
~~we should extract these into an `InMemoryStateInner{}` struct and then use an `Arc<InMemoryStateInner>` here~~